### PR TITLE
RUMM-2646 remove single GlobalRum monitor

### PR DESCRIPTION
**Pull request by xgouchet**
### What does this PR do?

A brief description of the change being made with this pull request.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: library/dd-sdk-android-rum/build.gradle.kts [hunk 1]
```diff
@@ -146,7 +146,7 @@ apply(from = "clone_telemetry_schema.gradle.kts")
 apply(from = "generate_rum_models.gradle.kts")
 apply(from = "generate_telemetry_models.gradle.kts")
 
-kotlinConfig()
+kotlinConfig(false)
```
Comment (0xnm) [inline library/dd-sdk-android-rum/build.gradle.kts:149]: we have some compilation warnings remaining?

File: library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/GlobalRum.kt [hunk 2]
```diff
@@ -27,40 +26,41 @@ import java.util.concurrent.atomic.AtomicBoolean
...
 
-    internal val globalAttributes: MutableMap<String, Any?> = ConcurrentHashMap()
-
-    internal val isRegistered = AtomicBoolean(false)
-    internal var monitor: RumMonitor = NoOpRumMonitor()
+    internal val registeredMonitors: MutableMap<SdkCore, RumMonitor> = mutableMapOf()
 
     /**
-     * Identify whether a [RumMonitor] has previously been registered.
+     * Identify whether a [RumMonitor] has previously been registered for the given Sdk instance.
```
Comment (0xnm) [inline library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/GlobalRum.kt:32]:
```suggestion
     * Identify whether a [RumMonitor] has previously been registered for the given SDK instance.
```

File: library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/RumFeature.kt [hunk 3]
```diff
@@ -142,6 +144,11 @@ class RumFeature internal constructor(
         backgroundEventTracking = configuration.backgroundEventTracking
         trackFrustrations = configuration.trackFrustrations
         rumEventMapper = configuration.rumEventMapper
```
Comment (0xnm) [inline library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/RumFeature.kt:146]: I think `rumEventMapper` is actually not used and can be removed, because when we initialize writer, we do it directly from `configuration.rumEventMapper` https://github.com/DataDog/dd-sdk-android/blob/3db8d5a9981b8dacfce9dcc49bdd8c5ced2604e4/library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/RumFeature.kt#L195

File: library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/event/RumEventMapper.kt [hunk 4]
```diff
@@ -32,6 +34,8 @@ internal data class RumEventMapper(
     val telemetryConfigurationMapper: EventMapper<TelemetryConfigurationEvent> = NoOpEventMapper()
 ) : EventMapper<Any> {
 
+    internal var sdkCore: SdkCore = NoOpSdkCore()
```
Comment (0xnm) [inline library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/event/RumEventMapper.kt:37]: we can inject it in the constructor, when we create a `RumDataWriter` in `RumFeature`

File: library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/monitor/DatadogRumMonitor.kt [hunk 5]
```diff
@@ -93,6 +93,8 @@ internal class DatadogRumMonitor(
         handler.postDelayed(keepAliveRunnable, KEEP_ALIVE_MS)
     }
 
+    private val globalAttributes = mutableMapOf<String, Any?>()
```
Comment (0xnm) [inline library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/monitor/DatadogRumMonitor.kt:96]: should we make it thread-safe?

File: library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/GlobalRum.kt [hunk 6]
```diff
@@ -27,40 +26,41 @@ import java.util.concurrent.atomic.AtomicBoolean
...
      *
+     * @param sdkCore the [SdkCore] instance to check against
      * @return whether a monitor has been registered
      * @see [registerIfAbsent]
      */
     @JvmStatic
-    fun isRegistered(): Boolean {
-        return isRegistered.get()
+    fun isRegistered(sdkCore: SdkCore): Boolean {
+        return synchronized(registeredMonitors) {
```
Comment (mariusc83) [inline library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/GlobalRum.kt:43]: why not using a `ConcurrentHashMap` instead ?

File: library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/instrumentation/UserActionTrackingStrategyLegacy.kt [hunk 7]
```diff
@@ -29,7 +29,7 @@ internal class UserActionTrackingStrategyLegacy(
 
     override fun onActivityResumed(activity: Activity) {
         super.onActivityResumed(activity)
-        gesturesTracker.startTracking(activity.window, activity)
+        sdkCore?.let { gesturesTracker.startTracking(activity.window, activity, it) }
```
Comment (0xnm) [inline library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/instrumentation/UserActionTrackingStrategyLegacy.kt:32]: how can we protect us from the situation when it won't be executed and we don't know about it?

**Review by 0xnm — COMMENTED** [review]
look good! I've added few comments to this draft.

Some changes regarding injection of `SdkCore` into some classes are pretty much the same as in my work for the logger, but this is ok, I will rebase later.

File: library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/GlobalRum.kt [hunk 8]
```diff
@@ -105,50 +102,29 @@ object GlobalRum {
...
-        globalAttributes.remove(key)
-    }
-
     // region Internal
 
-    internal fun notifyIngestedWebViewEvent() {
-        (monitor as? AdvancedRumMonitor)?.sendWebViewEvent()
+    internal fun clear() {
+        synchronized(registeredMonitors) {
+            registeredMonitors.clear()
```
Comment (mariusc83) [inline library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/GlobalRum.kt:119]: don't you need to call `clear` for each of the monitors here ?

File: library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/monitor/DatadogRumMonitor.kt (hunk 5 above)
Comment (mariusc83) [reply library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/monitor/DatadogRumMonitor.kt:96]: I think is lighter if we use a `ConcurrentHashMap` instead

File: build.gradle.kts [hunk 9]
```diff
@@ -268,7 +268,7 @@ tasks.register("buildNdkIntegrationTestsArtifacts") {
     dependsOn(":instrumented:integration:assembleDebug")
 }
 
-nightlyTestsCoverageConfig(threshold = 0.87f)
+nightlyTestsCoverageConfig(threshold = 0.7f)
```
Comment (0xnm) [inline build.gradle.kts:271]: signatures update for the API coverage will be handled in another PR?

**Review by 0xnm — COMMENTED** [review]
nice work! I reviewed most of it and left some minor comments, but then realized that it also overlaps with files/classes/methods which were changed in my PR for the internal logger, so it is better to rebase this PR.

File: library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/RumFeature.kt [hunk 10]
```diff
@@ -768,6 +771,8 @@ class RumFeature internal constructor(
         internal const val DD_TELEMETRY_CONFIG_SAMPLE_RATE_TAG =
             "_dd.telemetry.configuration_sample_rate"
 
+        internal val DEFAULT_NOOP_SDK_CORE = NoOpSdkCore()
```
Comment (0xnm) [inline library/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/RumFeature.kt:774]: I have a bit of concern regarding using `NoOpSdkCore` in features (and making it public). Normally features should respect lifecycle, so when we provide `NoOpSdkCore` if things go wrong they won't be noticed (because there is no logger behind).

But the other alternative will force us to either use `lateinit` or will add nullability. But at least if things go wrong we will get an exception (which normally we should get during our instrumentation/nightly tests, but there is a risk of exposure on the client side).

wdyt?

**Review by 0xnm — COMMENTED** [review]
nice! I've added few comments to discuss.

File: dd-sdk-android/src/main/kotlin/com/datadog/android/event/MapperSerializer.kt [hunk 11]
```diff
@@ -16,7 +16,7 @@ import com.datadog.android.core.persistence.Serializer
  * @param serializer Serializer to use.
  */
 class MapperSerializer<T : Any>(
-    internal val eventMapper: EventMapper<T>,
+    val eventMapper: EventMapper<T>,
```
Comment (0xnm) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/event/MapperSerializer.kt:19]: is this change needed?

File: library/dd-sdk-android-rum/src/test/kotlin/com/datadog/android/rum/ActivityViewTrackingStrategyTest.kt [hunk 12]
```diff
@@ -264,8 +266,9 @@ internal class ActivityViewTrackingStrategyTest :
     @Test
     fun `𝕄 start a RUM View event 𝕎 onActivityResumed()`() {
         // Given
+        testedStrategy.register(rumMonitor.mockSdkCore, mockAppContext)
         whenever(mockPredicate.accept(mockActivity)) doReturn true
-        testedStrategy.register(mockSdkCore, mockActivity)
+        testedStrategy.register(rumMonitor.mockSdkCore, mockActivity)
```
[the remark this answers is not in the document]
Comment (0xnm) [reply library/dd-sdk-android-rum/src/test/kotlin/com/datadog/android/rum/ActivityViewTrackingStrategyTest.kt:270]: seems like issue us still there

**Review by 0xnm — APPROVED** [review]
lgtm!

**Review by mariusc83 — APPROVED** [review]
Nice job !! LGTM
